Close #6525: linkcheck: Add linkcheck_ignore_redirects and linkcheck_warn_redirects

tk0miya opened the pull request:
### Feature or Bugfix
- Feature

### Purpose
- Add a new confval; `linkcheck_warn_redirects` to emit a warning when
    the hyperlink is redirected.  It's useful to detect unexpected redirects
    under the warn-is-error mode.
- Add a new confval; linkcheck_ignore_redirects to ignore hyperlinks
    that are redirected as expected.
- refs: #6525

tk0miya committed:
Close #6525: linkcheck: Add linkcheck_warn_redirects

Add a new confval; `linkcheck_warn_redirects` to emit a warning when
the hyperlink is redirected.  It's useful to detect unexpected redirects
under the warn-is-error mode.

tk0miya committed:
Close #6525: linkcheck: Add linkcheck_ignore_redirects

Add a new confval; linkcheck_ignore_redirects to ignore hyperlinks
that are redirected as expected.

tk0miya commented on sphinx/builders/linkcheck.py:
```diff
@@ -494,13 +498,23 @@ def check_uri() -> Tuple[str, str, int]:
-                if response.history:
+
+                if ignored_redirect(req_url, new_url):
+                    return 'working', '', 0
```
This should be marked as "ignored".

ngnpope replied:
I think that we want `'working'` and not `'ignored'` as, after all, we did check the link. `'ignored'` implies that we didn't even check.

The idea is that if it redirected to an expected target, we *don't* want it to show up as `'redirected'`. If it redirected to a target we didn't expect then we want it to show as `'redirected'` as before.

nomis commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,25 @@ Options for the linkcheck builder
 
    .. versionadded:: 1.1
 
+.. confval:: linkcheck_ignore_redirects
```
Rename this to `linkcheck_allowed_redirects`?

"Ignored" may imply that it's not followed and I think it should be checking the target of the redirect too.

ngnpope replied:
Yes. Sorry, that was my bad wording. I'm not sure that "allowed" is any better as that implies that other redirects are disallowed.

I'd suggest `linkcheck_silenced_redirects`. That makes clear that we're just silencing `'redirected'` and treating it as `'working'`.

tk0miya replied:
I feel "silenced" surely means suppressing warnings. But, I feel it does not mean changing `'redirected'` to `'working'`. So I prefer to use "allowed" instead of "ignore".

francoisfreitag commented on doc/usage/configuration.rst:
```diff
@@ -2642,6 +2661,14 @@ Options for the linkcheck builder
+
+   If true, emit a warning when redirection detected on checking hyperlinks.
+   It's useful to detect unexpected redirects under :option:`the warn-is-error
+   mode <sphinx-build -W>`.  Default is ``False``.
```
```suggestion
   If true, emits a warning when the response for a link is a redirect. It's
   useful to detect unexpected redirects under :option:`the warn-is-error mode
   <sphinx-build -W>`.  Default is ``False``.
```

francoisfreitag commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,25 @@ Options for the linkcheck builder
+   A dictionary that maps a pattern of the source URI to a pattern of the canonical
+   URI.  If set, linkcheck builder treated as "ignored" if a hyperlink in the source
+   document that matches to the pattern of the source URI, and it redirects to the
+   URL that matches to the pattern of the canonical URI.  Example:
```
```suggestion
   A dictionary that maps a source URI pattern to a canonical URI pattern. The
   linkcheck builder treats the redirected link as "ignored" when:

    - the link in the document matches the source URI pattern, and
    - the redirect location matches the canonical URI pattern.

   Example:
```

ngnpope replied:
Based on my other comments, we should change `"ignored"` → `"working"` and refer to "silencing" the redirect.

francoisfreitag commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,25 @@ Options for the linkcheck builder
+      }
+
+   It's helpful to enable :confval:`linkcheck_warn_redirects` not to warn the URIs
+   that causes HTTP redirection as expected.
```
```suggestion
   It's helpful to enable :confval:`linkcheck_warn_redirects` to warn for URIs
   causing unexpected HTTP redirection.
```

francoisfreitag commented on sphinx/builders/linkcheck.py:
```diff
@@ -494,13 +498,23 @@ def check_uri() -> Tuple[str, str, int]:
 
+        def ignored_redirect(url: str, new_url: str) -> bool:
+            for from_url, to_url in self.config.linkcheck_ignore_redirects.items():
+                if from_url.match(url) and to_url.match(new_url):
```
The specification from the issue also asks for a mean to reuse match groups from the `from_url` in the `to_url`.

IIUC, the issue asks something similar to:

```python
                from_match = from_url.match(url)
                if from_match:
                    to_url = from_match.expand(to_url)
                    if to_url.match(new_url):
                        return True
```

I am unsure if that is feasible at all.

From the [`sub()`](https://docs.python.org/3/library/re.html#re.sub) documentation (used by `expand()`):

> Unknown escapes of ASCII letters are reserved for future use and treated as errors.
For example:

```python
>>> import re
>>> to_url = r"^https://example.org/\w{2}/\1"
>>> match = re.match(r'https://example\.org/(.*)', 'https://example.org/the-real-page/')
>>> match.expand(to_url)
Traceback (most recent call last):
  File "/usr/lib/python3.9/sre_parse.py", line 1039, in parse_template
    this = chr(ESCAPES[this][1])
KeyError: '\\w'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/re.py", line 322, in _expand
    template = sre_parse.parse_template(template, pattern)
  File "/usr/lib/python3.9/sre_parse.py", line 1042, in parse_template
    raise s.error('bad escape %s' % this, len(this))
re.error: bad escape \w at position 21
```

There’s always the option of documenting that the canonical URI cannot contain special sequences (e.g. `\d`, `\w`, etc.), but that’s sure to bite most people.

---

Another idea to tighten the validation would be to use named groups in both the source URI and canonical URI, and compare the named groups. Something like:

```python
# conf.py
linkcheck_ignore_redirects = {
    'https://sphinx-doc.org/en/(?P<page>.*)': 'https://sphinx-doc.org/en/master/(?P<page>.*)',
}
# linkcheck
from_match = from_url.match(url)
to_match = to_url.match(to)
if from_match.groupdict() == to_match.group_dict():
    return True
```

That may not be as robust as one might expect. For example, I can imagine somebody documenting their regexp with capturing groups. IDK if the extra validation is worth the effort?

nomis replied:
If `https://www.sphinx-doc.org/en/stable/usage/quickstart.html` redirects to something other than `https://www.sphinx-doc.org/en/master/usage/quickstart.html` then it implies that the page no longer exists in the stable version and it should be possible to consider that an error. The validation should be worth the effort if someone is taking the effort to link to `/stable/` versions (although I'm not linking to readthedocs websites like this myself).

I would be inclined to use `re.sub()` and document it. Alternatively, do the group dict checking but only if the group name exists in both matches?

There may also need to be a regex and non-regex version because the need to escape every `.` for an accurate match may confuse people. Any examples should definitely use `r'...'` strings.

francoisfreitag replied:
> If https://www.sphinx-doc.org/en/stable/usage/quickstart.html redirects to something other than https://www.sphinx-doc.org/en/master/usage/quickstart.html then it implies that the page no longer exists in the stable version and it should be possible to consider that an error.

For a page that does not exist, the server should return a 404 in the first place.

I understand not all web servers behave that way, but let’s keep in mind we’re dealing with poorly configured web servers here. I wouldn’t advise implementing a brittle solution to work around bad behaving servers. I consider the `re.sub()` solution to be brittle:
- disallowing `\` in the `to_url` is sure to trip a good portion of users,
- `re.sub()` was design to perform replacements in a template string, not in a regular expression. There’s no guarantee the regular expression will be valid after the templating. The `\` is a symptom of it, but there may be others.

The comparison of groups seems more robust to me. One can adjust the groups to capture the same portion on both sides. It needs some work to experiment (e.g. try to implement the validation in the #6525 with that system). I’ll try it in the days to come.

> There may also need to be a regex and non-regex version because the need to escape every `.` for an accurate match may confuse people.

I think it would mostly be confusing. There would be two ways of doing the same thing. Besides, there’s a strong probability the unexpected redirect URL will be different enough that the regexp won’t match, even if `r"."` is used instead `r"\."`. Not discussing that `r"\."` is tighter and correct, simply trying to explain why a second setting without regexp seems overkill to me.

ngnpope replied:
As @nomis says, the idea here is that it is possible to accept redirect targets matching a certain pattern to be considered "working" and, if they don't match, then they should still show up as "redirected".

I've had a play with the example that @francoisfreitag posted and think that we can make this work by escaping non-backreference escapes:

```python
>>> import re
>>> to_url = r"^https://example.org/\w{2}/\1"
>>> match = re.match(r'https://example\.org/(.*)', 'https://example.org/the-real-page/')
>>> match.expand(to_url)
Traceback (most recent call last):
  File "/usr/lib/python3.9/sre_parse.py", line 1039, in parse_template
    this = chr(ESCAPES[this][1])
KeyError: '\\w'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/re.py", line 322, in _expand
    template = sre_parse.parse_template(template, pattern)
  File "/usr/lib/python3.9/sre_parse.py", line 1042, in parse_template
    raise s.error('bad escape %s' % this, len(this))
re.error: bad escape \w at position 21
>>> to_url2 = re.sub(r'(\\[^0-9g])', r'\\\1', to_url)
>>> to_url
'^https://example.org/\\w{2}/\\1'
>>> to_url2
'^https://example.org/\\\\w{2}/\\1'
>>> match.expand(to_url2)
'^https://example.org/\\w{2}/the-real-page/'
```

So `re.sub(r'(\\[^0-9g])', r'\\\1', to_url)` seems to do the trick, escaping non-backreference escapes. It seems that `.expand()` also reverses this extra escaping for us.

francoisfreitag commented on tests/test_build_linkcheck.py:
```diff
@@ -284,6 +286,36 @@ def test_follows_redirects_on_GET(app, capsys):
+    with http_server(make_redirect_handler(support_head=False)):
+        app.build()
+    assert ("index.rst.rst:1: WARNING: redirect  http://localhost:7777/ - with Found to "
+            "http://localhost:7777/?redirected=1\n" in strip_escseq(warning.getvalue()))
```
Perhaps verify this is the only warning:
```suggestion
    assert ("index.rst.rst:1: WARNING: redirect  http://localhost:7777/ - with Found to "
            "http://localhost:7777/?redirected=1\n" in strip_escseq(warning.getvalue()))
     assert warning.getvalue().count("\n") == 1

```

francoisfreitag reviewed (commented):
Overall the change looks good to me and is definitely an improvement. :+1:

---

Perhaps add two other tests, with a non-matching key and a non-matching value. In both tests, assert the redirection triggers a warning?

nomis commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,25 @@ Options for the linkcheck builder
+
+      linkcheck_ignore_redirects = {
+          # All HTTP redirections from the source URI to the canonical URI will be ignored.
+          'http://sphinx-doc.org/.*': 'https://sphinx-doc.org/en/master/.*'
```
I don't think the example `'https://sphinx-doc.org/.*': 'https://sphinx-doc.org/en/master/.*'` is good because it allows any redirect without checking that the page is actually the same.

For the specific case of `https://sphinx-doc.org/`, a redirect to anything in `https://(www\.)?sphinx-doc.org/.*` would be ok because the intent is just to link to the root page of the website.

nomis replied:
Since these are regexps it would have to be `r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/en/master/.*'` too...

ngnpope replied:
I'd make a small tweak to anchor it and drop the `.*`:

```suggestion
          r'^https?://sphinx-doc\.org/': r'^https://sphinx-doc\.org/en/master/'
```

(Obviously this will need updating if we need to use the capturing group approach.)

ngnpope reviewed (commented):
Thanks for taking this idea forward.

francoisfreitag commented:
> So re.sub(r'(\\[^0-9g])', r'\\\1', to_url) seems to do the trick, escaping non-backreference escapes. It seems that .expand() also reverses this extra escaping for us.

That seems to work in my testing :+1:, I had not thought of that option. It’s quite clean, with an identified set of patterns to look for.

I’m not yet convinced the approach of mangling the user regexp is valid and worth the complexity. Unknown parts come from the URL, the content would also need escaping. Consider the following example:

```python
>>> match = re.match(r'https://example\.org/(.*)', 'https://example.org/how-to-make-real-$$$/')
>>> match.expand(r"^https://example.org/\1")
'^https://example.org/how-to-make-real-$$$/'
>>> to_url = match.expand(r"^https://example.org/\1")
>>> re.match(to_url, 'https://example.org/how-to-make-real-$$$/')
```
Doesn’t match because `$` has a different semantic in the regexp.

It looks like applying `re.escape()` on all `match.groups()` would help, but I’m not sure how to re-inject that in the user regexp. 

A minor note, if we keep going down the route of crafting regular expressions, I think we should remove the regexp compilation step for the `to_url`.

Edit: Accidental send caused an intermediary version of the answer to be sent.

francoisfreitag commented:
The escaping approach prevents placing backrefs in the `to_url` regexp. Not sure if that is an issue in practice tough.

Example regexp:
```python
>>> import re
>>> re.match(r'https://example.org/([\w-]+)/\1/', 'https://example.org/foo/foo/')
<re.Match object; span=(0, 28), match='https://example.org/foo/foo/'>
>>> re.match(r'https://example.org/([\w-]+)/\1/', 'https://example.org/foo/bar/')
>>>
```

tk0miya committed:
Rename linkcheck_ignore_redirects to linkcheck_allowed_redirects

tk0miya committed:
Update docs

tk0miya committed:
Add testcase for combination of allow_redirects and warn_redirects

tk0miya commented:
I updated this PR except for the capturing feature. I think it's another topic. So I'd like to separate it to another PR (and it's helpful if somebody will post it :-)

nomis commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,28 @@ Options for the linkcheck builder
+
+      linkcheck_working_redirects = {
+          # All HTTP redirections from the source URI to the canonical URI will be treated as "working".
+          r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/en/master/.*'
```
If there's no regex capture then this isn't checking what it implies it's checking. To allow for a more complete example without comparing the source and destination paths, I've added an allowance for redirecting to `www.` but this could be removed too.

```suggestion
          r'https://sphinx-doc\.org/': r'https://(www\.)?sphinx-doc\.org/.*'
```

tk0miya replied:
I think the example should use patterns for both source and canonical URIs. So I think the current one is enough to describe the feature.

nomis replied:
But this is not just allowing redirects to the canonical version of a URL, it's allowing any redirects.

You may as well use `r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/.*'` because there isn't really any validation of the destination.

tk0miya replied:
What do you mean? The current rule verifies the pages under `https://sphinx-doc.org/` to the URL under `https://sphinx-doc.org/en/master/`.  It does not mean "any redirects".

nomis replied:
It's allowing practically any redirect, so not what I would consider a good example.

It's not far off allowing `https://www\.example\.com/.*` to `https://www\.example\.com/.*` which then causes linkcheck to ignore removed pages that just redirect to the home page.

tk0miya replied:
If `https://sphinx-doc.org/foo/bar` redirects to `https://sphinx-doc.org/other/path`, it's not "allowed". Is this allowing any redirect? I'm confused.

francoisfreitag commented on doc/usage/configuration.rst:
```diff
@@ -2522,6 +2522,28 @@ Options for the linkcheck builder
+
+   .. code-block:: python
+
+      linkcheck_working_redirects = {
```
```suggestion
      linkcheck_allowed_redirects = {
```

francoisfreitag reviewed (approved):
I believe this patch is an improvement to the linkcheck builder.

Before, it was not possible to treat redirects as errors. This patch offers the tools to do it. It also offers basic tools to verify the redirection occurs within the expected domain and path.

The redirection validation tools can be improved (e.g. by comparing match groups, or injecting backreferences captured from the source URI to the canonical URI). IMO, that should not prevent this first step from landing.

Thanks @tk0miya!

tk0miya committed:
Fix typo in doc/usage/configuration.rst

Co-authored-by: François Freitag <mail@franek.fr>

tk0miya committed:
Merge branch '4.x' into 6525_linkcheck_warn_redirects

tk0miya commented on doc/usage/configuration.rst:
```diff
@@ -2527,6 +2527,28 @@ Options for the linkcheck builder
+   URI. The linkcheck builder treats the redirected link as "working" when:
+
+    - the link in the document matches the source URI pattern, and
+    - the redirect location matches the canonical URI pattern.
```
Bad mark-up:

```suggestion
   - the link in the document matches the source URI pattern, and
   - the redirect location matches the canonical URI pattern.
```

tk0miya committed:
doc: Fix indentation

tk0miya commented on doc/usage/configuration.rst:
```diff
@@ -2647,6 +2669,14 @@ Options for the linkcheck builder
+   It's useful to detect unexpected redirects under :option:`the warn-is-error
+   mode <sphinx-build -W>`.  Default is ``False``.
+
+    .. versionadded:: 4.1
```
Bad mark-up too:

```suggestion
   .. versionadded:: 4.1
```

tk0miya committed:
doc: Fix indentation

tk0miya commented:
Now I'm debating to withdraw `linkcheck_warn_redirects`. After renaming to `linkcheck_allowed_redirects`, redirection that is not matched to the configuration is already considered as "disallowed". It's enough to emit a warning without `linkcheck_warn_redirects`.

I believe it's reasonable that the linkcheck builder emits warnings if "disallowed" redirection detected. In other words, it emits warnings if 1) `linkcheck_allowed_redirects` is configured and 2) a hyperlink having "redirected" status found.

tk0miya committed:
linkcheck: Emit a warning for disallowed redirects

Now linkcheck builder integrates `linkcheck_warn_redirects` into
`linkcheck_allowed_redirects`.  As a result, linkcheck builder will
emit a warning when "disallowed" redirection detected via
`linkcheck_allowed_redirects`.

tk0miya committed:
Merge branch '4.x' into 6525_linkcheck_warn_redirects

tk0miya commented:
It's time to release. I'm merging this. Please try the new option and let me know if something not working well.

Hopefully, it would be nice if somebody will send us a pull request that uses regexp as discussed above.
